### ISSUE TITLE
GetProgramExecutableName without loader

### ISSUE THREAD

**Issue by mrdomino** [jart/cosmopolitan#983]
As of #980 there is a happy path for `GetProgramExecutableName` with the ape loader installed.

This raises the question of what should happen if we are not on the happy path, e.g. the loader is not installed or is very out of date.

There are a few issues with the current (soon-to-be-fallback) code. For one, if it uses `_`, it doesn't absolutize it. For another, it doesn't work if the binary is configured as a login shell. (Wouldn't it be neat if you could use an ape binary as a login shell without the loader installed? Maybe not. But come on, how cool would that be.) There are other issues, those are just the ones I'm intimately familiar with.

As far as I can tell, the specs for this are:

- Performance is not critical, but it'd be nice for it to not be slow.
- We should preserve as much functionality sans loader as is reasonably possible.
- Code size should be kept to a minimum.

It's somewhat of an open question which things to try in what order. Current WIP draft of an answer (will be updated as discussion continues):

1. Try exotic methods like `/proc/self/exe` or `sysctl`. If that works, use that.
2. If `argv[0]` contains a slash and can be `access`'ed out of cwd, use that.
3. If `argv[0]` exists and contains no slash:
   1. If it is equal to `-$(basename $SHELL)`, use `$SHELL` or fail.
   2. Otherwise, if you can `commandv` it, use that.
4. If `_` exists:
   5. If it contains a slash and can be `access`'ed out of cwd, use that.
   6. Otherwise, if you can `commandv` it, use that.
5. Sad copy of `argv[0]` or `""`.

There is a question of whether to absolutize paths by simply prepending cwd or by `realpath()`.

This should all be disable-able with a "I only want this to run on systems with a loader" flag.

Thoughts?

Possibly related to #946.

**Comment by mrdomino** [comment]
I was actually thinking earlier it'd be really nice to just do `++envp` at some point in the initialization process. 😬

Where would be a good place to do that?

**Comment by jart** [comment]
What I'd do is add something like this to `libc/calls/getprogramexecutablename.greg.c`:

```c
__attribute((__constructor__)) static void init(int argc, char **argv, char **envp) {
  g_magicvar = *envp++;
}
```

**Comment by mrdomino** [comment]
Much of this was resolved by #1012. Much of the remainder is resolved by #1024. The rest is no longer relevant.
